src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

```diff
@@ -2123,6 +2156,123 @@ private void LoadEnvironmentVariable(ProcessStartInfo startinfo, IDictionary Env
             }
         }
 
+        /// <summary>
+        /// Attempt to stop the process when the timeout has expired.
+        /// <see cref="StopProcessCommand" /> is used to stop the process
+        /// </summary>
+        /// <param name="process">
+        /// The process that should be stopped
+        /// </param>
+        private void StopProcessOnTimeout(Process process)
+        {
+            string message = StringUtil.Format(ProcessResources.StartProcessTimeoutExceeded, process.ProcessName);
+            ErrorRecord er = new ErrorRecord(new TimeoutException(message), "StartProcessTimeoutExceeded", ErrorCategory.OperationTimeout, process);
+
+            StopProcessCommand stop = new StopProcessCommand();
+            stop.Id = GetProcessTreeIds(process);
+            foreach (Process p in stop.Invoke<Process>()) { }
+
+            ThrowTerminatingError(er);
+        }
+
+        /// <summary>
+        /// Gets IDs of descendant processes, started by a process
+        /// On Windows, this reads output from WMI commands
+        /// On UNIX, this reads output from `ps axo pid,ppid`
+        /// </summary>
+        /// <param name="parentProcess">
+        /// The parent process to use to resolve the process tree IDs
+        /// </param>
+        /// <returns>
+        /// IDs of the parent process and all its descendants
+        /// </returns>
+        private int[] GetProcessTreeIds(Process parentProcess)
+        {
+            List<int> stopProcessIds = new List<int> {parentProcess.Id};
+            string processRelationships = "";
+#if UNIX
+            try
+            {
+                Process ps = new Process();
+                ps.StartInfo.FileName = "ps";
+                ps.StartInfo.Arguments = "axo pid,ppid";
+                ps.StartInfo.UseShellExecute = false;
+                ps.StartInfo.RedirectStandardOutput = true;
+                ps.Start();
+
+                processRelationships = ps.StandardOutput.ReadToEnd();
+
+                ps.WaitForExit();
```

Comment:
Is there a potential this could not exit?

Reply:
Agreed. Shall add a few seconds wait before returning, just in case.